### PR TITLE
added applyVariables to Input fields

### PR DESCRIPTION
**Pull request by larkob**
This PR adds applyVariables to any fields used in the INPUT operation.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3214/pr-test (or any other room on that server)

**Comment by larkob** [comment]
I know the added code is pretty much a duplication, but I couldn't find a way to call a universal function because the `click` and `showInputOverlay` each are separate async functions with their own scope.

**Comment by UltimateGeek** [comment]
@larkob , could you update the pr-test room with an example of using this change?

**Comment by larkob** [comment]
Try to out the "Deal cards" button. It shows you your current playerName in the text input. You can then change it to some other name.

**Comment by larkob** [comment]
One minor issue I noticed is that a value parameter (the one with "Guest922244") is automatically added. No idea why that is. If I remove the line and edit the button again, it is automatically added.
```
    {
      "func": "INPUT",
      "header": "suggests name based on playerName variable",
      "fields": [
        {
          "type": "string",
          "label": "Your name?",
          "variable": "myName",
          "applyVariables": [
            {
              "parameter": "value",
              "variable": "playerName"
            }
          ],
          "value": "Guest922244"
        }
      ]
    }
```

**Comment by robartsd** [comment]
> One minor issue I noticed is that a value parameter (the one with "Guest922244") is automatically added. No idea why that is. If I remove the line and edit the button again, it is automatically added.
> 
> ```
>     {
>       "func": "INPUT",
>       "header": "suggests name based on playerName variable",
>       "fields": [
>         {
>           "type": "string",
>           "label": "Your name?",
>           "variable": "myName",
>           "applyVariables": [
>             {
>               "parameter": "value",
>               "variable": "playerName"
>             }
>           ],
>           "value": "Guest922244"
>         }
>       ]
>     }
> ```

This is what apply variables is doing. Variables in javascript don't actually contain objects of arrays, they contain a reference to the memory location where the object or array is stored. If you modify the object or array, all variables that reference that object or array will see the change; you don't get a completely different scope for the object or array unless you do a deep copy.

**Comment by pf-lkobbe** [comment]
I don't really understand deep vs. shallow copies. When I started on this, it seemed like a simple thing, but apparently it's not and so I'm not sure how to proceed with this PR (as it is clearly beyond my coding skills).

**Comment by larkob** [comment]
I don't really understand deep vs. shallow copies. When I started on this, it seemed like a simple thing, but apparently it's not and so I'm not sure how to proceed with this PR (as it is clearly beyond my coding skills).

File: client/js/widgets/button.js [hunk 1]
```diff
@@ -514,6 +514,27 @@ export class Button extends Widget {
...
+                field[v.parameter] = (v.index === undefined) ? variables[v.variable] : variables[v.variable][v.index];
+              } else if(v.parameter && v.template) {
+                field[v.parameter] = v.template.replace(/\{([^}]+)\}/g, function(i, key) {
+                  return (variables[key] === undefined) ? "" : variables[key];
+                });
+              } else if(v.parameter && v.property) {
+                let w = widgets.has(v.widget) ? widgets.get(v.widget) : this;
+                field[v.parameter] = (w.p(v.property) === undefined) ? null : w.p(v.property);
+              } else {
+                problems.push('Entry in parameter applyVariables does not contain "parameter" together with "variable", "property", or "template".');
```
Comment (UltimateGeek) [inline client/js/widgets/button.js:530]: I don't think there is a problems array in this context.

Comment (UltimateGeek) [reply client/js/widgets/button.js:530]: Take a look at my changes in #217 (which is pointed at this branch). I replicated the test in its pr-test room.

**Comment by rogerl50** [comment]
> > One minor issue I noticed is that a value parameter (the one with "Guest922244") is automatically added. No idea why that is. If I remove the line and edit the button again, it is automatically added.
> > ```
> >     {
> >       "func": "INPUT",
> >       "header": "suggests name based on playerName variable",
> >       "fields": [
> >         {
> >           "type": "string",
> >           "label": "Your name?",
> >           "variable": "myName",
> >           "applyVariables": [
> >             {
> >               "parameter": "value",
> >               "variable": "playerName"
> >             }
> >           ],
> >           "value": "Guest922244"
> >         }
> >       ]
> >     }
> > ```
> 
> This is what apply variables is doing. Variables in javascript don't actually contain objects or arrays, they contain a reference to the memory location where the object or array is stored. If you modify the object or array, all variables that reference that object or array will see the change; you don't get a completely different scope for the object or array unless you do a deep copy.
> 
> On line 69 `const a = { ...original }` performs a shallow copy of the routine function before proceeding. Keys of `a` that contain scalars have a copy of the scalar value that is not connected to the value in `original`; however, keys of `a` that contain objects contain the exact same object as `original` (not a copy). Thus applyVariables in the main function is not modifying `original` but applyVariables in the input fields is modifying `original`.
> 
> Perhaps line 69 should be changed to perform a deep copy instead of a shallow copy. Could also modify the applyVariables routine to work recursively so that the code is all in one place. If we did that, we'd need to decide if applyVariables works from the shallowest level to the deepest (more powerful as applyVariables could add an object that contains an applyVariables key; but could potentially cause an infinite loop) or from the deepest level to shallowest (safe, but less powerful).

That's really clear and interesting. Thanks. A question: would it be better to just do the deep copy once you are processing the `INPUT` tag? Doing it up top, in line 69, seems like a bit change. Are there any efficiency (or any other) concerns?

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
Looks fine. I agree with @UltimateGeek's comments.

- [ ] refactor `applyVariables`

**Review by UltimateGeek — APPROVED** [review]

**Review by RaphaelAlvez — APPROVED** [review]
tested and approved

**Comment by ArnoldSmith86** [comment]
Here are the save files of the PR server: [staging-PR-214.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/6093246/staging-PR-214.zip)

**Pull request merged**